src/neo/Network/P2P/RemoteNode.ProtocolHandler.cs

```diff
@@ -107,8 +108,17 @@ private void OnMessage(Message msg)
                     OnPongMessageReceived((PingPayload)msg.Payload);
                     break;
                 case MessageCommand.Transaction:
-                    if (msg.Payload.Size <= Transaction.MaxTransactionSize)
-                        OnInventoryReceived((Transaction)msg.Payload);
+                    Task.Run(() =>
+                    {
```

Comment:
A new task for each Transaction? I prefer to follow the same logic as OracleService, 4 tasks (or more) and consume a `BlockingCollection`

https://github.com/neo-project/neo/blob/d9c032b53aaeaa602d5a56f7bdfd424b83d6357b/src/neo/Oracle/OracleService.cs#L386-L402

Reply:
This need to be analyzed with careful as @shargon mentioned.

Reply:
ping @Qiao-Jin

Reply:
> ping @Qiao-Jin

I prefer restricting thread limit like what's done in #1666

Reply:
What it's the behaviour when it raise the max thread limit?

Reply:
> What it's the behaviour when it raise the max thread limit?

TPS will be low if max thread limit is too low or too high. Too much paralleled verification waiting in thread queue can affect other threads (akka actors), and affect overall performance

Reply:
Then I think that a fixed verification queues should be better. I will do it for compare the benchmarks.

Reply:
Please compare it with https://github.com/Qiao-Jin/neo/pull/39

Reply:
Hi @shargon , I tested Qiao-Jin#39 and compared with original code in my local env (a consensus node and a non-consensus one on 2 machines respectively) and got the following conclusion:

(1) numberOfTasks = 4 is a bit low. For example, consensus node can receive around 200000 tx during one block's period, but Transaction.VerifyIndependent is only performed around 70000 - 90000 times, and TPS is around 3000 - 3500.

(2) After changing numberOfTasks to some higher values (i.e.  10, 20) the problem is solved, TPS is enhanced (4500 - 5000 in case numberOfTasks = 10, 4000 - 4500 in case numberOfTasks = 20)

(3) In case of original code of this PR TPS is firmly around 5000 - 6000.

After observation the ability to perform Transaction.VerifyIndependent is the same, but the time used for persisting is different (I use #1666 in my local env)

Reply:
Did you check memory usage? Here, what happens if it overflows threads, its restricted by #1666 but transactions are waiting for thread slot or are discarded? perhaps it could lead to a denial of service if the transaction is waiting for a new thread slot. At https://github.com/Qiao-Jin/neo/pull/39 you have control over the number of transactions that you want in the queue.

Reply:
> Did you check memory usage?
Incoming akka messages will be queued into mailbox if not handled immediately. In this PR these messages are not queued in mailbox but made into tasks which are queued in system task queue. The memory usage is the same upon observation.

> what happens if it overflows threads, its restricted by #1666 but transactions are waiting for thread slot or are discarded?
Transactions are not discarded. Corresponding tasks are made & queued into system task queue.

> perhaps it could lead to a denial of service if the transaction is waiting for a new thread slot.
Denial of service will not occur. Say currently there are altogether 24 threads running and we cut max&min thread limit to 6. System will wait till there are less than 6 threads running. During this period new created tasks will be queued into system task queue. These queued tasks will be dequeued and run when there are less than 6 threads running.

> At Qiao-Jin#39 you have control over the number of transactions that you want in the queue
Yes, but other transactions will be still queued in actor mailbox. Memory used is the same.